Add support for Open Document Format 1.2

Naufragous opened the pull request:
Contains a kernel for the latest ODF 1.2 encryption implemented in LibreOffice. The algorithm uses a SHA-256 checksum, a PBKDF2-HMAC-SHA1 key derivation with 100000 iterations and key stretching and AES-CBC encryption.

Valid hashes can be extracted with the [libreoffice2john.py](https://github.com/magnumripper/JohnTheRipper/blob/bleeding-jumbo/run/libreoffice2john.py) script, available from the John the Ripper Jumbo repository.

Use hashcat with the `--username` option to process those hashes.

 - Add hash-mode 18400 (ODF 1.2 (SHA-256, PBKDF2-HMAC-SHA1, AES))
 - Tests: add hash-mode 18400 (ODF 1.2 (SHA-256, PBKDF2-HMAC-SHA1, AES))

Closes #101

jsteube commented:
Great work again, thanks!

Some comments:

- Please add a string substring (ODF) to the human readable descriptions for example in usage.c, interface.c, readme.txt and changes.txt. It's more likely people search for ODF instead of "Open Document Format"
- Please add missing entries in docs/readme 
- Please add missing entries in tab_completion/install.sh
- test.sh: Please add entry for 18400 to the SLOW_ALGOS array (it's possible there's some other modes missing as well)
- test.pl: The mode 18400 should not be listed in COMMON_UNSALTED_MODES - at least since this a salted algorithm it creates confusion with the word UNSALTED?! please check previous PR as well
- interface.c: This is ok since out_buf is a lot larger, but it still looks suspicious:

snprintf (&out_buf[159 + 2048], out_len - 1, "%s", odf12->suffix);

I'd have expected something like:

snprintf (&out_buf[159 + 2048], out_len - (1 + 159 + 2048), "%s", odf12->suffix);

but even that looks a bit strange. Maybe theres even a different solution with a temporary buffer

- interface.c:

missing hashconfig->hash_type      =  

- interface.h:

ROUNDS_LIBREOFFICE should be ROUNDS_ODF12 ?

- kernel:

please replace:

[16 * i + 0];

with:

const i16 = i * 16;

[i16 +  0] (notice 2x space because aligning)

- kernel:

please replace:

 pt1[0] = swap32_S (pt1[0] ^ iv[0]);

 with:

 pt1[0] ^= iv[0];

 Do all the swapping in a final block before hashing

Naufragous commented:
Thank you for the comments, I really did forget about a few things.
>add ODF to the human readable descriptions
>add missing entry in docs/readme.txt
>add missing entry in extra/tab_completion/hashcat.sh
>add 18400 to SLOW_ALGOS
>mode 18400 should not be listed in COMMON_UNSALTED_MODES
>interface.c: snprintf (&out_buf[...
>kernel: replace [16 * i + 0] with const i16
>kernel: group swaps together

All done.
>missing hashconfig->hash_type

Done, although not necessary for this hash mode.
>ROUNDS_LIBREOFFICE should be ROUNDS_ODF12?

Probably not... The ODF 1.2 standard does not require a specific iteration count beyond the 1,024 default. The parser expects anything between 1,000 and 999,999. The current LibreOffice version uses 100,000, that is why I choose it as the benchmark constant.

Updated the PR & rebased on current master (09ec2a9).

philsmd commented on tools/test.pl:
```diff
@@ -10597,6 +10598,43 @@ sub gen_hash
+  {
+    # Complex hash string with IV, salt, checksum, ciphertext and length
+    # Do not use salt_buf, do everything here in python
+    my $python_code = <<"END_CODE";
```
do we really need python code here?
I didn't find anything that can't be done with perl directly in the code block below. Maybe I'm missing something.
Please explain why you use python here

philsmd replied:
in general (a general question/discussion): it would also make sense to use a single additional programming language if perl really has absolutely no way/modules to do it. I think using PHP and python is acceptable, but if we can avoid it and stick to 1 additional language it would be even better. (I think the reason why we added the PHP code was just that perl had no modules and we needed that mode during a cmiyc contest etc... therefore it was a quick "hack")

Naufragous replied:
You are right, we don't need python code here. That's the just code I used while writing the kernel, and since I am not a perl monk I wrote it in python. But I have to rewrite this anyway, because right now there is no way to properly use the verify function.

Btw the last time I commited python code to the test suite was because of Streebog hashing. Didn't find anything on CPAN for that either.

philsmd replied:
yeah, I figured that out. so my idea would be to convert the code of -m 11900 = PBKDF2-HMAC-MD5 to python too such that we only have the python dependency. Do you think that makes sense? hashlib.pbkdf2_hmac in python2 probably ?
Furthermore, the base64 conversion of the words/salts (https://github.com/hashcat/hashcat/blob/master/tools/test.pl#L8226-L8227) should be done also for the python code. We do this to avoid any code injection and also to avoid any strange/special characters as input resulting in strange behaviour (this should be pretty easy to do in python too, just encode in perl and decode the strings in python).
we should also update tools/install_modules.sh if we need any additional/special python modules (btw: should we rename the file "install_modules.sh" into something that explains that this is needed for tests ? (install_test_modules.sh ? test_install_modules.sh ?) 

BTW: I do not think everything I mentioned here should be done within this commit/PR, the PHP->python could/should be a separate PR/issue (it's just something I noticed here and wanted to mention and ask if python is really the way to go here)

Naufragous replied:
Done! :D

philsmd commented on src/interface.c:
```diff
@@ -299,6 +299,7 @@ static const char *ST_HASH_18000 = "2fbf5c9080f0a704de2e915ba8fdae6ab00bbc026b2c
 static const char *ST_HASH_18100 = "597056:3600";
 static const char *ST_HASH_18200 = "$krb5asrep$23$user@domain.com:3e156ada591263b8aab0965f5aebd837$007497cb51b6c8116d6407a782ea0e1c5402b17db7afa6b05a6d30ed164a9933c754d720e279c6c573679bd27128fe77e5fea1f72334c1193c8ff0b370fadc6368bf2d49bbfdba4c5dccab95e8c8ebfdc75f438a0797dbfb2f8a1a5f4c423f9bfc1fea483342a11bd56a216f4d5158ccc4b224b52894fadfba3957dfe4b6b8f5f9f9fe422811a314768673e0c924340b8ccb84775ce9 ...
 static const char *ST_HASH_18300 = "$fvde$2$16$58778104701476542047675521040224$20000$39602e86b7cea4a34f4ff69ff6ed706d68954ee474de1d2a9f6a6f2d24d172001e484c1d4eaa237d";
+static const char *ST_HASH_18400 = "$odf$*1*1*100000*32*751854d8b90731ce0579f96bea6f0d4ac2fb2f546b31f1b6af9a5f66952a0bf4*16*2185a966155baa9e2fb597298febecbc*16*c18eaae34bcbbe9119be017fe5f8b52d*0*051e0f1ce0e866f2b771029e03a6c7119aad132af54c4e45824f16f61f357a40407ab82744fe6370c7b2346075fcd4c2e58ab244411b3ab1d532a46e2321599ef13c3d3472fc2f14d480d8c33215e473da67f90540279d3ef1f62dde314fa222796046e496c9 ...
```
What I noticed here is that you allow the file names to be included within the hash. As you can see here https://hashcat.net/wiki/example_hashes , we never include the file names within the hashes.
I think to be consistent, we also need to stick with -m 18400 to the standard format that hashcat doesn't use file names within the hashes (except if needed by the hashing algorithm). The problem is that if we add this "suffix" here, we would need to add it to many, many other formats too (like ms office, docx, pdf etc).
In my opinion the hash shouldn't be mixed with a file name (we could in theory use the --username feature to add some extra data/info at the beginning of the single lines of a hash file).

Naufragous replied:
That's what I initially did. Afterwards I added the suffix buffer to ensure a better UX; you can take the output of libreoffice2john.py and throw it at hashcat without editing. But if this is not required, then I actually would be glad to get rid of the suffix! Simplifies the parsing.

@jsteube, what will it be?

jsteube replied:
OK, remove it.

philsmd commented on src/interface.c:
```diff
@@ -10987,6 +10990,175 @@ int oldoffice34cm2_parse_hash (u8 *input_buf, u32 input_len, hash_t *hash_buf, M
+  token.attr[8]    = TOKEN_ATTR_VERIFY_LENGTH
+                   | TOKEN_ATTR_VERIFY_DIGIT;
+
+  token.len_min[9] = 32;
```
hmm, shouldn't these len_min == len_max all using TOKEN_ATTR_FIXED_LENGTH instead ?

Naufragous replied:
I'd be happy to use `FIXED_LENGTH` (like I did for the 2048 chars of ciphertext at the end), but the tokenizer will automatically omit the separating `*` only if I use `VERIFY_LENGTH`. Or I am missing something?

jsteube replied:
The TOKEN_ATTR_FIXED_LENGTH is ideal for hashes when there is no separator character (like descrypt) while at the same time there are still different columns. The macro name maybe is not ideal to reflect that, but TOKEN_ATTR_VERIFY_LENGTH is correct use here.

Naufragous commented:
Update:
- use perl modules instead of python code in the test suite
- complete test suite with proper verify and passthrough branches
- remove support for hash suffix generated by libreoffice2john.py


Rebased on current master (0db6913).

Naufragous committed:
Add support for Open Document Format 1.2

Contains a kernel for the latest ODF 1.2 encryption implemented in
LibreOffice. The algorithm uses a SHA-256 checksum, a PBKDF2-HMAC-SHA1
key derivation with 100000 iterations and key stretching and AES-CBC
encryption.

Valid hashes can be extracted with the libreoffice2john.py script,
available from the John the Ripper Jumbo repository at
https://github.com/magnumripper/JohnTheRipper/blob/bleeding-jumbo/run/libreoffice2john.py

You have to remove the filename suffix at the end of the hash before
passing it to hashcat. Also see 'hashcat -m18400 --example-hashes'.

You can leave the filename prefix if you use the --username option to
process those hashes.

 - Add hash-mode 18400 (Open Document Format (ODF) 1.2 (SHA-256, AES))
 - Tests: add hash-mode 18400 (Open Document Format (ODF) 1.2 (SHA-256, AES))

philsmd commented on docs/changes.txt:
```diff
@@ -1,11 +1,23 @@
+## Technical
+##
+
+- Tests: Added hash-mode 18400 (Open Document Format (ODF) 1.2 (SHA-256, AES))
```
minor consideration: we normally do not add any lines/entries for new tests in docs/changes.txt (except if the tests were missing in older releases and we finally managed to complete/add the tests). I think it is enough to just have the other line under Algorithms, because usually if we add a new algo it comes accompanied with tests (as in this case). Again, this is not really a problem. It's just a note and a hint to make it more consistent for other/future PRs ;)

jsteube commented:
Many thanks!
